packages/http/libraries/core/src/adapter/InversifyHttpAdapter.ts

```diff
@@ -0,0 +1,209 @@
+import { Container } from '@inversifyjs/container';
+import { getReflectMetadata } from '@inversifyjs/reflect-metadata-utils';
+
+import { ControllerMetadata } from '../models/ControllerMetadata';
+import { ControllerMethodMetadata } from '../models/ControllerMethodMetadata';
+import { ControllerMethodParameterMetadata } from '../models/ControllerMethodParameterMetadata';
+import { METADATA_KEY } from '../models/MetadataKey';
+import { RequestMethodParameterType } from '../models/RequestMethodParameterType';
+import { RouterParams } from '../models/RouterParams';
+
+export abstract class InversifyHttpAdapter<
+  TRequest,
+  TResponse,
+  TNextFunction extends (err?: unknown) => void,
+> {
+  readonly #container: Container;
+
+  constructor(container: Container) {
+    this.#container = container;
+  }
+
+  protected _buildServer(): void {
+    this.#registerControllers();
+  }
+
+  #registerControllers(): void {
+    const controllerMetadataList: ControllerMetadata[] | undefined =
+      getReflectMetadata(Reflect, METADATA_KEY.controller);
+
+    if (controllerMetadataList === undefined) {
+      throw new Error('No controllers found');
+    }
+
+    this.#buildHandlers(controllerMetadataList);
+  }
+
+  #buildHandlers(controllerMetadataList: ControllerMetadata[]): void {
+    for (const controllerMetadata of controllerMetadataList) {
+      const controllerMethodMetadataList:
+        | ControllerMethodMetadata[]
+        | undefined = getReflectMetadata(
+        controllerMetadata.target,
+        METADATA_KEY.controllerMethod,
+      );
+
+      if (controllerMethodMetadataList !== undefined) {
+        const routerParams: RouterParams<TRequest, TResponse, TNextFunction>[] =
+          this.#buildRouterParams(
+            controllerMetadata,
+            controllerMethodMetadataList,
+          );
+        this._buildRouter(controllerMetadata.path, routerParams);
+      }
+    }
+  }
+
+  #buildRouterParams(
+    controllerMetadata: ControllerMetadata,
+    controllerMethodMetadataList: ControllerMethodMetadata[],
+  ): RouterParams<TRequest, TResponse, TNextFunction>[] {
+    return controllerMethodMetadataList.map(
+      (controllerMethodMetadata: ControllerMethodMetadata) => {
+        const parameterMetadata:
+          | {
+              [key: string]: ControllerMethodParameterMetadata[];
+            }
+          | undefined = getReflectMetadata(
+          controllerMetadata.target,
+          METADATA_KEY.controllerMethodParameter,
+        );
+
+        const controller: { [key: string | symbol]: () => unknown } =
+          this.#container.get(controllerMetadata.target);
+
+        return {
+          handler: async (
+            req: TRequest,
+            res: TResponse,
+            next: TNextFunction,
```

Comment:
I wonder if this approach is generic enough, I believe so but maybe we should double check

Reply:
It's not, but imho this is a first iteration